CART-565 hlc: Initial Hybrid Logical Clocks implementation

dmiter opened the pull request:
Inserted HLC timestamps to all RPCs send and received.
This allow to arrange them according the time across all nodes
involved in communication. Disregard to RPCs you can query for
current timestamp like following:

uint64_t hlc = crt_hlc_get();

This function returns the current HLC timestamp for this moment.

daosbuild1 commented on src/test/test_hlc_net.c:
```diff
@@ -0,0 +1,315 @@
+	((uint32_t)		(src)			CRT_VAR) \
+	((uint32_t)		(dst)			CRT_VAR)
+
+#define CRT_OSEQ_RPC_TEST	/* output fields */		 \
```
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 commented on src/test/test_hlc_net.c:
```diff
@@ -0,0 +1,315 @@
+#define DEBUG		1
+#define MAX_SEQ		1000
+
+#define CRT_ISEQ_RPC_TEST	/* input fields */		 \
```
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/1/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/1/display/redirect

daosbuild1 commented:
Test stage Build on Leap 15 with Intel-C completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/1/display/redirect

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/3/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/3/display/redirect

daosbuild1 commented:
Test stage Build on Leap 15 with Intel-C completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/2/display/redirect

daosbuild1 commented:
Test stage Build on Leap 15 with Intel-C completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/3/display/redirect

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/4/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/4/display/redirect

daosbuild1 commented:
Test stage Build on Leap 15 with Intel-C completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/4/display/redirect

daosbuild1 commented on src/cart/crt_hlc.c:
```diff
@@ -0,0 +1,120 @@
+#endif
+#include <mercury_atomic.h>
+#ifdef __INTEL_COMPILER
+#pragma warning (pop)
```
(style)  space prohibited between function name and open parenthesis '('

daosbuild1 replied:
(style)  space prohibited between function name and open parenthesis '('

daosbuild1 commented on src/cart/crt_hlc.c:
```diff
@@ -0,0 +1,120 @@
+ */
+#include "crt_hlc.h"
+#ifdef __INTEL_COMPILER
+#pragma warning (push)
```
(style)  space prohibited between function name and open parenthesis '('

daosbuild1 replied:
(style)  space prohibited between function name and open parenthesis '('

daosbuild1 commented on src/cart/crt_hlc.c:
```diff
@@ -0,0 +1,120 @@
+#include "crt_hlc.h"
+#ifdef __INTEL_COMPILER
+#pragma warning (push)
+#pragma warning (disable: 2330)
```
(style)  space prohibited between function name and open parenthesis '('

(style)  spaces required around that ':' (ctx:VxW)

daosbuild1 replied:
(style)  space prohibited between function name and open parenthesis '('

(style)  spaces required around that ':' (ctx:VxW)

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/5/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/5/display/redirect

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/6/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/6/display/redirect

daosbuild1 commented on src/test/test_hlc_net.c:
```diff
@@ -0,0 +1,323 @@
+	((uint32_t)		(src)			CRT_VAR) \
+	((uint32_t)		(dst)			CRT_VAR)
+
+#define CRT_OSEQ_RPC_TEST	/* output fields */		 \
```
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 commented on src/test/test_hlc_net.c:
```diff
@@ -0,0 +1,323 @@
+#define DEBUG		1
+#define MAX_SEQ		1000
+
+#define CRT_ISEQ_RPC_TEST	/* input fields */		 \
```
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 commented on src/cart/crt_hlc.c:
```diff
@@ -0,0 +1,126 @@
+#endif
+#include <mercury_atomic.h>
+#ifdef __INTEL_COMPILER
+#pragma warning (pop)
```
(style)  space prohibited between function name and open parenthesis '('

daosbuild1 commented on src/cart/crt_hlc.c:
```diff
@@ -0,0 +1,126 @@
+ */
+#include "crt_hlc.h"
+#ifdef __INTEL_COMPILER
+#pragma warning (push)
```
(style)  space prohibited between function name and open parenthesis '('

daosbuild1 commented on src/cart/crt_hlc.c:
```diff
@@ -0,0 +1,126 @@
+#include "crt_hlc.h"
+#ifdef __INTEL_COMPILER
+#pragma warning (push)
+#pragma warning (disable: 2330)
```
(style)  space prohibited between function name and open parenthesis '('

(style)  spaces required around that ':' (ctx:VxW)

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/7/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/7/display/redirect

jolivier23 commented:
I wonder if we shouldn’t just move iof_atomic.h to gurt.  It has a cas and wouldn’t require openpa

johannlombardi commented:
> I wonder if we shouldn’t just move iof_atomic.h to gurt. It has a cas and wouldn’t require openpa

Agreed.

johannlombardi commented on src/cart/crt_hlc.c:
```diff
@@ -0,0 +1,126 @@
+
+static hg_atomic_int64_t crt_hlc = HG_ATOMIC_VAR_INIT(0);
+
+static inline int crt_hlc_gettime(uint64_t *time)
```
Should probably just be crt_hlc_get().

dmiter replied:
This is internal function. I have a public crt_hlc_get() defined at the end of this file.

johannlombardi commented on src/cart/crt_hlc.c:
```diff
@@ -0,0 +1,126 @@
+
+	return 0;
+}
+
```
I don't think we have any use case for this function.

johannlombardi commented on src/cart/crt_hlc.c:
```diff
@@ -0,0 +1,126 @@
+
+	return 0;
+}
+
```
Same a above

ashleypittman replied:
A comment explaining what this function does and why it's used would be useful.

johannlombardi commented on src/cart/crt_hlc.c:
```diff
@@ -0,0 +1,126 @@
+
+	return 0;
+}
+
```
hm, i am not sure I understand the API actually. For me, we need at most two functions: one to get the current time and once to get the time when a request was received. That's it. The rest should be internal glue not visible to the cart user.

dmiter replied:
We have them. They declared in api.h

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/8/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/8/display/redirect

dmiter commented:
> > I wonder if we shouldn’t just move iof_atomic.h to gurt. It has a cas and wouldn’t require openpa
> Agreed.
Done.

daosbuild1 commented:
Test stage Build on Leap 15 with Intel-C completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/8/display/redirect

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/9/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/9/display/redirect

daosbuild1 commented:
Test stage Build on Ubuntu 18.04 with Clang completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/9/display/redirect

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/10/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/10/display/redirect

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/11/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/11/display/redirect

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/12/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/12/display/redirect

johannlombardi commented:
Could you guys please prioritize this review? Thanks in advance.

ashleypittman commented on src/cart/crt_hlc.c:
```diff
@@ -0,0 +1,128 @@
+	hlc = rc ? atomic_load_consume(&crt_hlc)
+		 : (now.tv_sec * 1e9 + now.tv_nsec);
+
+	return hlc & ~CRT_HLC_MASK;
```
If clock_gettime() fails do we want to mask in the HLC_MASK?

ashleypittman commented on src/cart/crt_hlc.c:
```diff
@@ -0,0 +1,128 @@
+	uint64_t pt = crt_hlc_localtime_get();
+
+	do {
+		hlc = atomic_load_consume(&crt_hlc);
```
Same comment above about atomics.

jolivier23 replied:
My point is that I think it's a noop on intel but I could be wrong.

ashleypittman commented on src/cart/crt_rpc.c:
```diff
@@ -1304,6 +1304,30 @@ crt_req_abort(crt_rpc_t *req)
+{
+	struct crt_rpc_priv	*rpc_priv;
+
+	D_ASSERT(req != NULL);
```
If this is a public API then it should return -DER_INVAL here.

ashleypittman commented on src/cart/crt_rpc.c:
```diff
@@ -1304,6 +1304,30 @@ crt_req_abort(crt_rpc_t *req)
+{
+	struct crt_rpc_priv	*rpc_priv;
+
+	D_ASSERT(req != NULL);
```
As above.

ashleypittman commented on src/include/cart/api.h:
```diff
@@ -407,6 +407,34 @@ crt_reply_get(crt_rpc_t *rpc)
 }
 
+/**
+ * Return HLC timestamp of RPC was sent
```
This comment needs extending, the timestamp of the req_send() function?  What does this mean for RPCs which are being sent, vs RPCs which are being received?

dmiter replied:
done.

ashleypittman commented on src/include/cart/api.h:
```diff
@@ -407,6 +407,34 @@ crt_reply_get(crt_rpc_t *rpc)
+crt_hlc_send_get(crt_rpc_t *req);
+
+/**
+ * Return HLC timestamp of RPC (new or reply) was received
```
What does "received" mean in this context?  Is this only to be called on the Tx side and received is when the reply was received on the node or is it something else?

dmiter replied:
done

ashleypittman commented on src/include/cart/api.h:
```diff
@@ -407,6 +407,34 @@ crt_reply_get(crt_rpc_t *rpc)
+crt_hlc_receive_get(crt_rpc_t *req);
+
+/**
+ * Return HLC timestamp of current event
```
current time?

dmiter replied:
yes

ashleypittman reviewed (commented):
What happens with HLC when process sets/groups attach to each other?  Is there the idea of a local vs a remote clock?

jolivier23 commented on src/cart/crt_hlc.c:
```diff
@@ -0,0 +1,128 @@
+
+	rc = clock_gettime(CLOCK_REALTIME_COARSE, &now);
+	hlc = rc ? atomic_load_consume(&crt_hlc)
+		 : (now.tv_sec * 1e9 + now.tv_nsec);
```
you could use NSEC_PER_SEC rather than constant (defined in gurt/common.h)

jolivier23 commented on src/include/cart/api.h:
```diff
@@ -407,6 +407,34 @@ crt_reply_get(crt_rpc_t *rpc)
+ * \return                     HLC timestamp
+ */
+uint64_t
+crt_hlc_get(void);
```
In my opinion, the public API should simply be int crt_hlc_get(const crt_rpc_t *rpc);

And all of the send/receive should be handled internally to cart.   The public API would either grab the latest RPC time if you pass non-NULL or the latest time if you don't.

These are automatically attached to RPCs anyway, no?  The user doesn't need to know that.

ashleypittman replied:
Alternatively, given we have a public `struct crt_rpc` we should just add them there and let the user read them directly, adding a function to fetch a private struct entry doesn't add anything other than overhead.

dmiter replied:
I specially split API into several functions. For RPC we have two different timestamps. In case of rpc == NULL which timestamp I should return, send or receive?

I am not in favor of introducing two more fields into the public structure. This will increase the memory footprint and reduce the flexibility of possibility to ​​change them later.

jolivier23 replied:
Regardless, it still doesn't answer my question.   I'm not seeing the value of send/receive timestamps.  I can see the value of getting last rpc timestamp.

dmiter replied:
Disagree. I think this is very useful. If it will not be used I remove this.

jolivier23 replied:
It's obvious you disagree as you haven't changed it.   But what is the use case you have in mind?  The reason we want HLC in DAOS is purely to approximate a perfectly synchronized global clock for the purpose of timestamping updates to DAOS objects.  For that, we need the latest timestamp.  I can see the utility of the latest timestamp associated with an RPC received in order to avoid excessive calls to clock_gettime but I just don't see DAOS using the timestamp of the send for anything but perhaps logging and CART can probably handle that internally on rpc send/receive.

ashleypittman replied:
Doing something that's based on NTP for synchronization isn't going to be any use to DAOS is it?

The self-test code was originally intended to report on-the-wire times for RPCs, so it could be update to measure from the "send" timestamp, depending on the exact definition.

dmiter replied:
Ok. I remove this stuff. So, all as you wish.

jolivier23 replied:
I didn't mean to remove the capability to retrieve the hlc associated with a message but rather just to remove the differentiation between send/receive.   It would still be useful to avoid the clock_getttime when assigning a timestamp to an I/O by using that latest timestamp associated with an RPC.

Johann had suggested something like crt_hlc_get(crt_rpc_t *rpc); where if rpc == NULL, we get a new timestamp, otherwise, use the latest one associated with the rpc.   Though I think Ashley's suggestion to have it as a public field in the public part of the data structure may be cleaner.   We could simply keep a copy of the value in the common header that is updated on changes.

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/13/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/13/display/redirect

liuxuezhao commented on src/cart/crt_hg_proc.c:
```diff
@@ -722,15 +734,19 @@ crt_proc_in_common(crt_proc_t proc, crt_rpc_input_t *data)
 			D_GOTO(out, rc);
 		}
-		if (proc_op == CRT_PROC_DECODE)
-			rpc_priv->crp_flags = rpc_priv->crp_req_hdr.cch_flags;
```
hmm, why remove above 2 lines?
it was used when forward corpc from parent to child if I remember correctly.
I'll request a change for this.

dmiter replied:
From my point of view this is dead code path. I add the following in the beginning of this function:
D_ASSERT(proc_op != CRT_PROC_DECODE);
and then run applications. I didn't get this assert. This function is used only for ENCODE.

Could you point me exact case where it's used in DECODE?

liuxuezhao replied:
I see why did not see the DECODE case here.
crt_proc_in_common is the proc func registered to HG, it will be called in two paths:
1) within HG_Forward -> hg_set_input ..., to pack in the request
2) when received the RPC, if user calls HG_Get_input it will call it with DECODE. We used to call HG_Get_input very early phase so that was needed. And I changed that handling to be crt_hg_unpack_header + _unpack_body and removed the direct calling of HG_Get_input. 
So you are right that no DECODE case in our current usage. But for code integrity, I think it is better to keep the above two lines (it will cause bug if we change to directly call HG_Get_input some time later), or comment out it with some additional comments to make it clear.

dmiter replied:
Thanks for explanation. I'd better comment this and put assert in case of logic change.

liuxuezhao commented on src/cart/crt_rpc.c:
```diff
@@ -1304,6 +1304,30 @@ crt_req_abort(crt_rpc_t *req)
+}
+
+uint64_t
+crt_hlc_receive_get(crt_rpc_t *req)
```
probably can check crp_state, for example if the RPC is still in flight then the returned hlc is non-sense. (return error or just and a log or err msg?)
or in API comment explicitly say returned hlc is un-trustable if the RPC is still inflight.

dmiter replied:
My assumption is you will use this API in completion callback of RPC. So, you will have an information about RPC status. I'm not in favor of block this call until RPC completion or check for status.

I also assume there are no errors can be in this call but it can return zero instead of correct HLC timestamp. From my point of view this is fine because of HLC should be used to provide correct order. Zero timestamp is always less all other timestamps.

liuxuezhao commented on src/cart/crt_hlc.c:
```diff
@@ -0,0 +1,110 @@
+	uint64_t pt = crt_hlc_localtime_get();
+
+	do {
+		hlc = atomic_load_consume(&crt_hlc);
```
just curious that can't we just use the OPA atomic API?

dmiter replied:
:-) I'm just avoid of using them by Ashley request and introduce atomics in gurt for this.

jolivier23 replied:
It was my request.   Mainly because opa adds an external dependency (though one that we already build).   Given that the language is providing them, I don't see much value in adding the opa directly.

liuxuezhao reviewed (changes requested):
the patch generally looks fine to me.
the request change just for the "rpc_priv->crp_flags" change (or maybe I miss understood sth).

BTW, can you consolidated the following commits in this PR in one patch? that seems more convenient to review.

dmiter commented on src/cart/crt_hlc.c:
```diff
@@ -0,0 +1,110 @@
+	uint64_t	hlc;
+	int		rc;
+
+	rc = clock_gettime(CLOCK_REALTIME_COARSE, &now);
```
Be monotonic is not critical for HLC algorithm. But NTP can help in precise of time synchronization. So, I need a real physical clock of current node.

dmiter commented on src/cart/crt_hlc.c:
```diff
@@ -0,0 +1,110 @@
+#include <gurt/atomic.h>
+#include <time.h>
+
+#define CRT_HLC_MASK 0xFFFF
```
ok.

dmiter commented on src/cart/crt_hlc.c:
```diff
@@ -0,0 +1,110 @@
+	do {
+		hlc = atomic_load_consume(&crt_hlc);
+		new = (hlc & ~CRT_HLC_MASK) < pt ? pt : (hlc + 1);
+	} while (!atomic_compare_exchange(&crt_hlc, hlc, new));
```
yes, the HLC should be more for each next call. So, next call will return HLC+1 at least.

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/14/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/14/display/redirect

liuxuezhao reviewed (dismissed):
one point in my mind is the change of:
if (proc_op == CRT_PROC_DECODE)
			rpc_priv->crp_flags = rpc_priv->crp_req_hdr.cch_flags;

removing the code breaks the code integrity, as the crt_proc_in_common should with that handing, the logic is incorrect without it, although that path is not called in our usage. If you want to remove it better to add comments.
but it needs not to block the patch.

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/15/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/15/display/redirect

gnailzenh commented on src/cart/crt_hg_proc.c:
```diff
@@ -541,6 +546,8 @@ crt_hg_unpack_header(hg_handle_t handle, struct crt_rpc_priv *rpc_priv,
 		D_GOTO(out, rc);
 	}
+	rpc_priv->crp_reply_hdr.cch_hlc = crt_hlc_get_msg(
+						rpc_priv->crp_req_hdr.cch_hlc);
```
so this member is being reused to store the received HLC, so later call of crt_hlc_receive_get() can extract it from this message right?

The HLC of the outgoing reply will be overwritten while packing the reply?

liuxuezhao replied:
I think here it is just to update local crt_htc, the assignment to "crp_reply_hdr.cch_hlc" seems useless here, as later when sending the reply in crt_proc_out_common it will re-assign it: crp_reply_hdr.cch_hlc = crt_hlc_get();

dmiter replied:
Here I update a local HLC and store receiving timestamp which will be get by crt_hlc_receive_get(rpc). This is receive new RPC code path. For replys it will be set in different place (@liuxuezhao point to this case). But this is not overriding. This is two different code paths.

dmiter replied:
Hmm. I need to refine this.

dmiter replied:
reworked this.

dmiter commented on src/cart/crt_hlc.c:
```diff
@@ -0,0 +1,117 @@
+	int		rc;
+
+	rc = clock_gettime(CLOCK_REALTIME_COARSE, &now);
+	pt = rc ? atomic_load_consume(&crt_hlc)
```
I observed few fails in my local testing. This error affected correct ordering. So, I resolved this. In case of error I just use previous time.

jolivier23 replied:
That seems error prone.   What if previous time is a long time ago?

gnailzenh replied:
Based on man page, the only two errors could be returned are, EINVAL and EPERM, the first one is about invalid parameters, so I'm wondering how it could fail in your case

dmiter replied:
I didn't remember exact error. Probably it was EBUSY. Anyway if I use old time value it works stable.

ashleypittman commented on src/cart/crt_hlc.c:
```diff
@@ -0,0 +1,128 @@
+	int		rc;
+
+	rc = clock_gettime(CLOCK_REALTIME_COARSE, &now);
+	hlc = rc ? atomic_load_consume(&crt_hlc)
```
That's not what I was getting at, it's nothing to do with values across CPUs.  atomic_load_consume() will cause the local CPU to not reorder any reads/writes across the atomic access, effectively stopping and restarting the CPU pipeline which will come with a cost.  As we are just counting here, and not doing any synchronization we should use the version of the atomics that doesn't prevent reordering

Commit:
CART-565 hlc: Initial Hybrid Logical Clocks implementation

Inserted HLC timestamps to all RPCs send and received.
This allow to arrange them according the time across all nodes
involved in communication. Disregard to RPCs you can query for
current timestamp like following:

uint64_t hlc = crt_hlc_get();

This function returns the current HLC timestamp for this moment.

Signed-off-by: Dmitry Eremin <dmitry.eremin@intel.com>

daosbuild1 commented on src/test/test_hlc_net.c:
```diff
@@ -0,0 +1,310 @@
+	((uint32_t)		(src)			CRT_VAR) \
+	((uint32_t)		(dst)			CRT_VAR)
+
+#define CRT_OSEQ_RPC_TEST	/* output fields */		 \
```
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 commented on src/test/test_hlc_net.c:
```diff
@@ -0,0 +1,310 @@
+#define DEBUG		1
+#define MAX_SEQ		1000
+
+#define CRT_ISEQ_RPC_TEST	/* input fields */		 \
```
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 replied:
(style)  Macros with complex values should be enclosed in parentheses

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/16/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/16/display/redirect

ashleypittman commented on src/cart/crt_hlc.h:
```diff
@@ -0,0 +1,50 @@
+#include <stdint.h>
+
+/* Internal API to sync timestamp with remote message */
+uint64_t crt_hlc_get_msg(uint64_t msg);
```
This seems like it could go in crt_internal_fns.h and avoid having a new file at all.

dmiter replied:
ok

ashleypittman commented on src/cart/crt_hlc.c:
```diff
@@ -0,0 +1,128 @@
+ * incompatible with parameter of type "const volatile void *"
+ * (dropping qualifiers)
+ */
+#pragma warning(disable:2330)
```
Yulu was looking at this, but I think he fixed it so it may just be case of merging with master.

dmiter replied:
He resolved a different issue.

ashleypittman commented on src/cart/crt_hg_proc.c:
```diff
@@ -541,6 +546,7 @@ crt_hg_unpack_header(hg_handle_t handle, struct crt_rpc_priv *rpc_priv,
 		D_ERROR("crt_proc_common_hdr failed rc: %d.\n", rc);
 		D_GOTO(out, rc);
 	}
+	(void)crt_hlc_get_msg(rpc_priv->crp_req_hdr.cch_hlc);
```
Why are you casting to (void) here?  Is this code just to increase the clock but not to store the value?

dmiter replied:
yes, It syncs and drop HLC result of syncing returned from this functions.

ashleypittman commented on src/cart/crt_hg_proc.c:
```diff
@@ -781,12 +800,16 @@ crt_proc_out_common(crt_proc_t proc, crt_rpc_output_t *data)
 
 	if (proc_op != CRT_PROC_FREE) {
+		if (proc_op == CRT_PROC_ENCODE)
+			rpc_priv->crp_reply_hdr.cch_hlc = crt_hlc_get();
```
So this is recording when the RPC was replied to, and putting in it the header?  Would it be better to call crt_hlc_get() in crt_reply_send() and just store the previous value here?

dmiter replied:
This function will be called right before sending reply into the wire (When Mercury will encode the packet). So, I think this is right place.

ashleypittman replied:
This is the kind of information that should be in the header file for the public API, that the RPC send timestamp is the point in time that the RPC was transmitted from the local node.

dmiter replied:
done

ashleypittman reviewed (commented):
I'm confused about this local patch, it doesn't seem to export the timestamps for RPCs at all any more?

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/17/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/17/display/redirect

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/18/display/redirect

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/19/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/19/display/redirect

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/20/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/20/display/redirect

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/21/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/21/display/redirect

jolivier23 reviewed (commented):
If we decide to expose a field, I think it can land as is and that can be done in a separate change.  If we decide to add optional RPC argument, I think it should be changed before landing.

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/22/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/22/display/redirect

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/25/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/25/display/redirect

ashleypittman committed:
Use memory_order_relaxed for atomic operations.

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/26/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/26/display/redirect

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/27/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/27/display/redirect

dmiter commented:
All tests passed
02:13:32  [Pipeline] End of Pipeline
02:13:32  [BFA] Scanning build for known causes...
02:13:32  [BFA] No failure causes found
02:13:32  [BFA] Done. 0s
02:13:31  sh: line 1: 22796 Terminated              sleep 3
02:13:32  GitHub has been notified of this commit’s build result
02:13:32  Finished: FAILURE

I don't understand what the issue. Maybe just timeout...

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/28/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/cart/job/PR-73/28/display/redirect

Nasf-Fan commented on src/cart/crt_hlc.c:
```diff
@@ -0,0 +1,98 @@
+#include <gurt/atomic.h>
+#include <time.h>
+
+#define CRT_HLC_MASK 0xFFFFULL
```
For 48-bits of physical time, the mask should be "0x0FFFULL" or I missed anything?

dmiter replied:
The physical time is most significant 48 bits. So, this mask is the rest of it. 64 - 48 = 16
